kong/pdk/service/response.lua

```diff
@@ -25,21 +25,36 @@ local header_body_log = phase_checker.new(PHASES.header_filter,
                                           PHASES.log)
 
 
-local function headers(response_headers, err)
-  local mt = getmetatable(response_headers)
-  local index = mt.__index
-  mt.__index = function(_, name)
+local attach_mt_index
+
+
+do
+  local resp_headers_orig_mt_index
```

Comment:
Ahh, so all returns of ngx.resp.get_headers in a Lua state share the same metatable then? Good catch! That explains the leak, as the chain of functions is formed. I mistakenly assumed both the returned table and metatable would be owned by the caller (then the original code wouldn't leak).

Since the metatable is unique, I guess it would be a better idea to produce a new table here, rather than patch the existing one, otherwise we're changing the semantics of `ngx.resp.get_headers`.

Reply:
I have considered that, but find this pre-existing approach more efficient.

Reply:
We can create our own metatable and reuse it, essentially keeping the efficiency. This will prevent surprises in scenarios mixing PDK and raw-ngx plugins.

Reply:
Which example do you have in mind that is currently broken between the two?

Reply:
If some code in a plugin is using raw ngx.resp.get_headers and does not expect the additional logic that we add via the metatable to be there.

Reply:
Oh, gotcha. Yeah of course, that makes sense. Somehow yesterday night I overlooked that this overrides the ngx_lua metatable...